### PR TITLE
feat!: Update default values for security posture improvement

### LINKED ISSUE #324: Fix: RDS Read Replica - Do not create random password module is created as a read replica
https://github.com/terraform-aws-modules/terraform-aws-rds/blob/a853247232927c6ec62c9264c45cd377f899ce74/main.tf#L12-L17

This isn't really an error. Wanted to cleanup where when this module is created for a read replica we don't create a hanging random password resource

### LINKED ISSUE #354: Tfsec says there is a critical vulnerability in variables.tf
## Description

The latest release of tfsec (0.58.4) identifies (perhaps incorrectly) this repo as having a couple of issues:
```
  Result 1

  [general-secrets-sensitive-in-variable][CRITICAL] Variable 'variable.password' includes a potentially sensitive default value.
  terraform-aws-rds/variables.tf:129


     126 | variable "password" {
     127 |   description = "Password for the master DB user. Note that this may show up in logs, and it will be stored in the state file"
     128 |   type        = string
     129 |   default     = ""    string: ""
     130 | }
     131 | 
     132 | variable "port" {

  Legacy ID:  GEN001
  Impact:     Default values could be exposing sensitive data
  Resolution: Don't include sensitive data in variable defaults

  More Info:
  - https://tfsec.dev/docs/general/secrets/sensitive-in-variable#general/secrets 
  - https://www.terraform.io/docs/state/sensitive-data.html 

  Result 2

  [aws-rds-encrypt-instance-storage-data][HIGH] Resource 'module.db_instance:aws_db_instance.this[0]' has storage encrypted set to false
  terraform-aws-rds/modules/db_instance/main.tf:30


      27 |   instance_class    = var.instance_class
      28 |   allocated_storage = var.allocated_storage
      29 |   storage_type      = var.storage_type
      30 |   storage_encrypted = var.storage_encrypted    bool: false
      31 |   kms_key_id        = var.kms_key_id
      32 |   license_model     = var.license_model
      33 | 

  Legacy ID:  AWS052
  Impact:     Data can be read from RDS instances if compromised
  Resolution: Enable encryption for RDS instances

  More Info:
  - https://tfsec.dev/docs/aws/rds/encrypt-instance-storage-data#aws/rds 
  - https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance 
  - https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.Encryption.html 

  Result 3

  [general-secrets-sensitive-in-local][CRITICAL] Local 'locals.' includes a potentially sensitive value which is defined within the project.
  terraform-aws-rds/main.tf:2


       1 | locals {
       2 |   master_password      = var.create_db_instance && var.create_random_password ? random_password.master_password[0].result : var.password    string: ""
       3 |   db_subnet_group_name = !var.cross_region_replica && var.replicate_source_db != null ? null : coalesce(var.db_subnet_group_name, module.db_subnet_group.db_subnet_group_id, var.identifier)
       4 | 
       5 |   parameter_group_name_id = var.create_db_parameter_group ? module.db_parameter_group.db_parameter_group_id : var.parameter_group_name

  Legacy ID:  GEN002
  Impact:     Local value could be leaking secrets
  Resolution: Don't include sensitive data in locals

  More Info:
  - https://tfsec.dev/docs/general/secrets/sensitive-in-local#general/secrets 
  - https://www.terraform.io/docs/state/sensitive-data.html 
```

Some of these may be false-positives in which case `#tfsec:ignore/XXXXX` comment tags could be added. But if not then can these be addressed?

## Versions
Terraform: 1.0.0
 - Provider(s):
provider.aws v3.37.0

## Reproduction

 - Clone the repo: `git clone https://github.com/terraform-aws-modules/terraform-aws-rds.git`
 - Download the latest tfsec, e.g. `wget https://github.com/liamg/tfsec/releases/latest/download/tfsec-linux-amd64 -O tfsec`
 - Run `./tfsec`

### LINKED ISSUE #362: restore_to_point_in_time for RDS SQL Server instance engine
## Is your request related to a new offering from AWS?
Restore to Point In Time seems supported by RDS SQL Server since initial release.

## Describe the solution you'd like.
This module can easily support Restore to Point In Time for AWS RDS SQL Server Instance supported feature as do for not mssql Intance `false == local.is_mssql`

This feature is key for RDS SQL Server Instance Disaster Recovery strategy.

### LINKED ISSUE #373: Oracle DB ReplicaMode Support
## Is your request related to a new offering from AWS?

https://github.com/hashicorp/terraform-provider-aws/issues/17991
TF has since added support for the Oracle RDS replica mode (`mounted` or `open-read-only` with default of `open-read-only`)

## Describe the solution you'd like.

It would be great to have an optional flag for this!

## Describe alternatives you've considered.

Since we cannot utilize the replica mode, we are unable to use this module and have to create our own for Oracle. 😿

### LINKED ISSUE #377: "engine" conflicts with "replicate_source_db" in hashicorp/aws 4.0.0
When trying to create a MySQL read replica, in "hashicorp/aws 4.0.0", "engine" conflicts with "replicate_source_db" since its default value is "". It should be set as "null". However, in that case, Terraform throws an error because of "is_mssql" local definition ("engine" should not be "null).

`│ Error: Conflicting configuration arguments
│ 
│   with module.db.module.db_instance.aws_db_instance.this[0],
│   on .terraform/modules/db/modules/db_instance/main.tf line 25, in resource "aws_db_instance" "this":
│   25:   engine            = var.engine
│ 
│ "engine": conflicts with replicate_source_db`

`│ Error: Invalid function argument
│ 
│   on .terraform/modules/db/modules/db_instance/main.tf line 2, in locals:
│    2:   is_mssql = element(split("-", var.engine), 0) == "sqlserver"
│     ├────────────────
│     │ var.engine is null
│ 
│ Invalid value for "str" parameter: argument must not be null.`

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- update outputs to use `try()` function
- update default settings for passwords to fix security settings
	- do not set an empty string as default (ref #354)
	- change default behavior to use randomly generated password of length 16 (was 10 previously)
- update default settings to enable encryption by default (ref #354)
- set `username`, `password`, and `engine` to `null` when  a `replicate_source_db` or `snapshot_identifier` is provided; these values are already provided in the respective source (ref #324)
- add new attribute `replica_mode` and bump AWS provider version to support (ref #357)
- add `latest_restorable_time` to ignored changes (ref #356)
- correct `db_subnet_group_name` to use full name when prefix is enabled (ref #359)
- add `restore_to_point_in_time` attribute to MSSQL instance (ref #363)
- bump min supported version of Terraform to 0.13.1
- bump random provider to be consistent using v4.x due to `name` being deprecated and `db_name` replacing it

For complete list of changes, please see upgrade guide https://github.com/bryantbiggs/terraform-aws-rds/blob/feat/updates/UPGRADE-4.0.md

## Motivation and Context

- Closes #105
- Closes #324
- Closes #354
- Closes #356
- Closes #357
- Closes #359 
- Closes #362
- Closes #363 
- Closes #373
- Closes #374
- Closes #377

## Breaking Changes
- Yes

## How Has This Been Tested?
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

### CODE REVIEW

**Comment by bryantbiggs** [comment]
@antonbabenko ps - it looks like their might be an issue with the token or something on the PR title check for this repo, not sure `Error: Resource not accessible by integration`

File: variables.tf [hunk 1]
```diff
@@ -121,12 +121,14 @@ variable "username" {
...
   default     = null
+  sensitive   = true
 }
 
 variable "password" {
   description = "Password for the master DB user. Note that this may show up in logs, and it will be stored in the state file"
   type        = string
-  default     = ""
+  default     = null
+  sensitive   = true
```
Comment (antonbabenko) [inline variables.tf:131]: `sensitive` was added in Terraform 0.14. I would not use this feature still in the module.

Comment (bryantbiggs) [reply variables.tf:131]: ah right, mixing up my inputs and outputs. reverted from the variables section

File: examples/complete-oracle/main.tf [hunk 2]
```diff
@@ -64,22 +64,19 @@ module "db" {
   identifier = "demodb-oracle"
 
   engine               = "oracle-ee"
-  engine_version       = "12.1.0.2.v8"
-  family               = "oracle-ee-12.1" # DB parameter group
-  major_engine_version = "12.1"           # DB option group
+  engine_version       = "19.0.0.0.ru-2021-10.rur-2021-10.r1"
+  family               = "oracle-ee-19.0" # DB parameter group
+  major_engine_version = "19.0"           # DB option group
```
Comment (antonbabenko) [inline examples/complete-oracle/main.tf:69]: What an upgrade! From 12 to 19 :)

File: examples/s3-import-mysql/main.tf [hunk 3]
```diff
@@ -80,17 +80,25 @@ module "security_group" {
   tags = local.tags
 }
 
-module "import_s3_bucket" {
-  source  = "terraform-aws-modules/s3-bucket/aws"
-  version = "~> 2.0"
-
+# Temporary work around until S3 module is updated to support v4.x
```
Comment (antonbabenko) [inline examples/s3-import-mysql/main.tf:83]: FYI I am planning to work on the S3 bucket module on Friday and probably next week.

**Comment by antonbabenko** [comment]
This PR is included in version 4.0.0 :tada:

**Pull request merged**